_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,321 @@
+
+[Ray]:https://docs.ray.io
+[Ray Tune]:https://docs.ray.io/en/master/tune.html
+
+[dml-ms]:https://ml.dask.org/hyper-parameter-search.html
+[skl-ms]:https://scikit-learn.org/stable/modules/grid_search.html
+[tsne]:https://distill.pub/2016/misread-tsne/
+[tuning]:https://scikit-learn.org/stable/modules/grid_search.html
+
+[Ridge]:https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html#sklearn.linear_model.Ridge
+[1]:https://medium.com/distributed-computing-with-ray/gridsearchcv-2-0-new-and-improved-ee56644cbabf
+
+> Cutting edge hyperparameter tuning techniques (bayesian optimization, early
+> stopping, distributed execution) can provide significant speedups over grid
+> search and random search.
+>
+> However, the machine learning ecosystem is missing a solution that provides
+> users with the ability to leverage these new algorithms while allowing users
+> to stay within the Scikit-Learn API. In this blog post, we introduce
+> tune-sklearn [Ray's tuning library] to bridge this gap. Tune-sklearn is a
+> drop-in replacement for Scikit-Learn’s model selection module with
+> state-of-the-art optimization features.
+>
+> —[GridSearchCV 2.0 — New and Improved][1]
+
+This claim is inaccurate: Dask-ML has provided access to "cutting edge
+hyperparameter tuning techniques" with a Scikit-learn compatible API for over a
+year. To provide correct this inaccuracy, let's start by looking at the
+features of Ray's tune-sklearn:
+
+[dml-hod]:https://ml.dask.org/hyper-parameter-search.html
+
+> Here’s what [Ray's] tune-sklearn has to offer:
+> 1. **Consistency with Scikit-Learn API** ...
+> 2. **Modern hyperparameter tuning techniques** ...
+> 3. **Framework support** ...
+> 4. **Scale up** ...
+>
+> [Ray's] Tune-sklearn is also **fast**.
+
+Dask-ML's model selection module has every one of the features. Let's look at
+each feature and see what Dask-ML and Ray have to offer:
+
+> 1\. **Consistency with Scikit-Learn API**: tune-sklearn is a drop-in
+> replacement for GridSearchCV and RandomizedSearchCV, so you only need to
+> change less than 5 lines in a standard Scikit-Learn script to use the API.
+
+*Dask-ML is also consistent with the Scikit-learn API.*
+
+Here's how to use Dask-ML's hyperparameter optimization:
+
+``` python
+from dask_ml.model_selection import HyperbandSearchCV
+from dask.distributed import Client
+from scipy.stats import loguniform
+from sklearn.datasets import make_classification
+from sklearn.linear_model import SGDClassifier
+
+if __name__ == "__main__":
+    X, y = make_classification()
+    model = SGDClassifier()
+    params = {"loss": ["hinge", "log", "modified_huber", "squared_hinge", "perceptron"],
+              "alpha": loguniform(1e-5, 1e-3)}
+
+    client = Client()
+    search = HyperbandSearchCV(model, params)
+    search.fit(X, y, classes=[0, 1])
+```
+
+This example is almost identical to [one of Ray's examples][sgd-eg] and [one of
```

Comment:
I know the formatting might be hard, but I think side by side code blocks would be the best way to demonstrate this. Maybe you could use a table?

Reply:
Yeah, formatting is difficult. You're saying side-by-side with Scikit-learn?

Reply:
side-by-side dask-ml and ray? Or ideally both ray and sk-learn

Reply:
How's that? I did "top-and-bottom" for Dask-ML, Ray and Scikit-learn instead of "side-by-side":

``` python
from scipy.stats import loguniform
from sklearn.datasets import make_classification
from sklearn.linear_model import SGDClassifier
from dask.distributed import Client
client = Client()

X, y = make_classification()
model = SGDClassifier()
params = {"loss": ["hinge", "log", "modified_huber", "squared_hinge", "perceptron"],
          "alpha": loguniform(1e-5, 1e-3)}

# Scikit-learn hyperparameter optimization
from sklearn.model_selection import RandomizedSearchCV
search = RandomizedSearchCV(clf, params, n_iter=20)
search.fit(X, y)

# Dask hyperparameter optimization
from dask_ml.model_selection import HyperbandSearchCV
search = HyperbandSearchCV(model, params)
search.fit(X, y, classes=[0, 1])

# Ray hyperparameter optimization
from tune_sklearn import TuneSearchCV
search = TuneSearchCV(clf, params, early_stopping=True)
search.fit(X, y, classes=[0, 1])
```